go/common/grpc/proxy/proxy.go

```diff
@@ -0,0 +1,229 @@
+	// between all requets.
+}
+
+// TODO: this is duplicated with code in grpc/policy. Extract and reuse.
+func extractPeerFromContext(ctx context.Context) (string, error) {
+	peer, ok := peer.FromContext(ctx)
+	if !ok {
+		return "", errors.New("grpc: failed to obtain connection peer from context")
+	}
+	tlsAuth, ok := peer.AuthInfo.(credentials.TLSInfo)
+	if !ok {
+		return "", errors.New("grpc: unexpected peer authentication credentials")
+	}
+	if nPeerCerts := len(tlsAuth.State.PeerCertificates); nPeerCerts != 1 {
+		return "", fmt.Errorf("grpc: unexpected number of peer certificates: %d", nPeerCerts)
+	}
+	peerCert := tlsAuth.State.PeerCertificates[0]
+	subject := accessctl.SubjectFromX509Certificate(peerCert)
+	return string(subject), nil
+}
+
+func (p *proxy) handler(srv interface{}, stream grpc.ServerStream) error {
+	method, ok := grpc.MethodFromServerStream(stream)
+	if !ok {
+		p.logger.Error("missing method in client request")
+		return status.Errorf(codes.Internal, "missing method in client request")
+	}
+
+	// Upstream stream.
+	upstreamCtx, upstreamCancel := context.WithCancel(stream.Context())
+	defer upstreamCancel()
+	desc := &grpc.StreamDesc{
+		ServerStreams: true,
+		ClientStreams: true,
+	}
+	sub, err := extractPeerFromContext(upstreamCtx)
+	if err != nil {
+		p.logger.Error("failed extracting peer from context",
+			"err", err,
+		)
+		return status.Errorf(codes.Internal, "failed extracting peer from context")
+	}
+	// Pass subject header upstream.
+	upstreamCtx = metadata.AppendToOutgoingContext(upstreamCtx, policy.ForwardedSubjectMD, sub)
+
+	upstreamStream, err := grpc.NewClientStream(
+		upstreamCtx,
+		desc,
+		p.upstreamConn,
+		method,
+	)
+
+	if err != nil {
+		return err
+	}
+
+	// Proxy upstream.
+	upErrCh := p.proxyUpstream(stream, upstreamStream)
+
+	// Proxy downstream.
+	downErrCh := p.proxyDownstream(upstreamStream, stream)
+
+	// Wait for errors.
+	for {
+		select {
+		case err := <-upErrCh:
+			if err == io.EOF {
+				// Received EOF from downstream client, the server response
+				// can still be in progress.
+				p.logger.Debug("downstream EOF")
+				if err = upstreamStream.CloseSend(); err != nil {
+					p.logger.Error("failrue closing upstream stream",
+						"err", err,
+					)
+				}
+			} else {
+				return err
+			}
+			break
+		case err := <-downErrCh:
+			// Received Error from upstream server.
+			// Proxy trailer if remains and propagate the error.
+			stream.SetTrailer(upstreamStream.Trailer())
+			if err == io.EOF {
+				p.logger.Debug("upstream EOF")
+				return nil
+			}
+			return err
+		}
+	}
+}
+
+// Client -> Upstream
+func (p *proxy) proxyUpstream(downstream grpc.ServerStream, upstream grpc.ClientStream) <-chan error {
+	errCh := make(chan error, 1)
+
+	go func() {
+		for {
+			// XXX: since we are using CBOR we are able to unmarshal messages
+			// without knowing the schema. This wouldn't work with protobuf, and
+			// a raw binary codec should be used instead.
+			var m interface{}
```

Comment:
You can probably use `cbor.RawMessage` instead of `interface{}` to avoid any deserialization from taking place.